server/ast/select.go

```diff
@@ -0,0 +1,206 @@
+		selectStmt.Limit = limit
+		return selectStmt, nil
+	case *vitess.SetOp:
+		selectStmt.OrderBy = orderBy
+		selectStmt.With = with
+		selectStmt.Limit = limit
+		return selectStmt, nil
+	default:
+		return nil, fmt.Errorf("SELECT has encountered an unknown clause: `%T`", selectStmt)
+	}
+}
+
+// nodeSelectStatement handles tree.SelectStatement nodes.
+func nodeSelectStatement(node tree.SelectStatement) (vitess.SelectStatement, error) {
+	if node == nil {
+		return nil, nil
+	}
+	switch node := node.(type) {
+	case *tree.ParenSelect:
+		return nodeParenSelect(node)
+	case *tree.SelectClause:
+		return nodeSelectClause(node)
+	case *tree.UnionClause:
+		return nodeUnionClause(node)
+	case *tree.ValuesClause:
+		return nodeValuesClause(node)
+	default:
+		return nil, fmt.Errorf("unknown type of SELECT statement: `%T`", node)
+	}
+}
+
+// nodeSelectExpr handles tree.SelectExpr nodes.
+func nodeSelectExpr(node tree.SelectExpr) (vitess.SelectExpr, error) {
+	switch expr := node.Expr.(type) {
+	case *tree.AllColumnsSelector:
+		if expr.TableName.NumParts > 1 {
+			return nil, fmt.Errorf("referencing items outside the schema or database is not yet supported")
+		}
+		return &vitess.StarExpr{
+			TableName: vitess.TableName{
+				Name: vitess.NewTableIdent(expr.TableName.Parts[0]),
+			},
+		}, nil
+	case tree.UnqualifiedStar:
+		return &vitess.StarExpr{}, nil
+	case *tree.UnresolvedName:
+		if expr.NumParts > 2 {
+			return nil, fmt.Errorf("referencing items outside the schema or database is not yet supported")
+		}
+		if expr.Star {
+			var tableName vitess.TableName
+			if expr.NumParts == 2 {
+				tableName.Name = vitess.NewTableIdent(expr.Parts[1])
+			}
+			return &vitess.StarExpr{
+				TableName: tableName,
+			}, nil
+		} else {
+			var tableName vitess.TableName
+			if expr.NumParts == 2 {
+				tableName.Name = vitess.NewTableIdent(expr.Parts[1])
+			}
+			return &vitess.AliasedExpr{
+				Expr: &vitess.ColName{
+					Name:      vitess.NewColIdent(expr.Parts[0]),
+					Qualifier: tableName,
+				},
+				As:              vitess.NewColIdent(string(node.As)),
+				InputExpression: tree.AsString(&node),
+			}, nil
+		}
+	default:
+		vitessExpr, err := nodeExpr(expr)
+		if err != nil {
+			return nil, err
+		}
+		return &vitess.AliasedExpr{
+			Expr:            vitessExpr,
+			As:              vitess.NewColIdent(string(node.As)),
+			InputExpression: tree.AsString(&node),
+		}, nil
+	}
+}
+
+// nodeSelectExprs handles tree.SelectExprs nodes.
+func nodeSelectExprs(node tree.SelectExprs) (vitess.SelectExprs, error) {
+	if len(node) == 0 {
+		return nil, nil
+	}
+	selectExprs := make(vitess.SelectExprs, len(node))
+	for i := range node {
+		var err error
+		selectExprs[i], err = nodeSelectExpr(node[i])
+		if err != nil {
+			return nil, err
+		}
+	}
+	return selectExprs, nil
+}
+
+// nodeExprToSelectExpr handles tree.Expr nodes and returns the result as a vitess.SelectExpr.
+func nodeExprToSelectExpr(node tree.Expr) (vitess.SelectExpr, error) {
```

Comment:
Unused

Reply:
This is used by `nodeExprsToSelectExprs` right under it, and `nodeExprsToSelectExprs` is used in other files. I always make a distinction between the single and collective variants, for the same reasons mentioned elsewhere in the PR. I don't want someone to later make `nodeExprToSelectExpr` when the logic should already be consolidated in one place.